drivers/led_strip/ws2812_uart.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2024 Croxel, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT worldsemi_ws2812_uart
+
+#include <zephyr/drivers/led_strip.h>
+
+#include <string.h>
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_uart);
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/math_extras.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/dt-bindings/led/led.h>
+
+struct ws2812_uart_cfg {
+	const struct device *uart_dev; /* UART device */
+	uint8_t *px_buf;
+	size_t length;
+	uint8_t num_colors;
+	const uint8_t *color_mapping;
+	const uint8_t *uart_lut;
+	uint8_t bits_per_rgb_bit;
+	uint16_t reset_delay;
+};
+
+/*
+ * Serialize a 24-bit color channel value into an equivalent sequence
+ * of 8 bytes of uart frames,.
+ */
+static inline int ws2812_uart_ser(const struct device *dev, uint8_t *buf, uint8_t color[3])
+{
+	/* Explaination of serialisation for WS2812 ic:
+	 * 3 pulses/bits of UART are used to represent one bit of RGB data.
+	 * During the development of this driver, I was using an ESP32C3 MCU
+	 * with its UART baud rate set to 2.5Mbits per sec.Hence,1 bit equals 400ns.
+	 * For WS2812 timing details, please refer to its datasheet.
+	 * Link: https://cdn-shop.adafruit.com/datasheets/WS2812.pdf
+	 * We can represent logic 1 with 0b110 (first two bits high,
+	 * remaining one bit low),
+	 * and logic 0 with 0b100 (first bit high, remaining two bits low).
+	 * We also need to invert the polarity of UART pins because in default,
+	 * the idle TX line is high, but WS2812 or similar LEDs need the idle
+	 * line to be low, otherwise, it will misinterpret.
+	 * When we make the TX line inverted, the start bit will be low to high
+	 * and the stop bit will be high to low.
+	 * Hence, the nine bits we are sending in one go will always have 1 at
+	 * the start and 0 at the end, we just need to calculate middle 7 bits.
+	 * Also, keep in mind that UART always sends the MSB first.
+	 * Keeping all this data in mind, I created a lookup table to ease the
+	 * calculations.
+	 * This function is just taking three bits at a time from the 24-bit color
+	 * data, i.e., color[3] array (MSB first),
+	 * and passing those three bits into the lookup table ws2812_uart_lut[] to
+	 * retrieve the 7-bit data and fills out
+	 * the serialized output buffer.
+	 * This driver can be used with other ic of different timing requirements
+	 * by simply changing baudrate, bits per rgb bit and lookup table.
+	 */
+
+	int i, j;
+	int buff_indx = 0;
+	uint8_t serialized_val = 0;
+	const struct ws2812_uart_cfg *cfg = dev->config;
+
+
+	for (i = 0; i < 3; i++) {
+		for (j = 0; j < 8; j++) {
+			serialized_val = (serialized_val << 1) | ((color[i]>>(7-j))&0x01);
+			if (((i*8)+j) % cfg->bits_per_rgb_bit == cfg->bits_per_rgb_bit-1) {
```

Comment:
Is it correct ?

Let's say we have `bits-per-rgb-bit=<4>`, `data-bits=<6>` and 4 entries in the `lookup-table` (representing all the possible values for 2 RGB bits). So we should pick-up a value in the `lookup-table` every two RGB bits. But with the above line it is every 4 RGB bits... Am I wrong ?

Reply:
Yeah you are right. Thanks for picking this up. I have updated this.